### DISCUSSION TITLE
projection of strain increment on the flow direction

### DISCUSSION THREAD

**Discussion by GhiathMonnet** [thelfer/tfel#378] [Q&A]
Hi every body.
I would like to know how much of isotropic plastic strain (dp) is contained in the increment of the total strain deto. In other words, I need to compute dp from deto. I guess I must project (deto) on the flow direction (n). Is there a dedicated tool for this operation in mfront  ? Usually we do compute the plastic strain amount in deto as : dp*n. I would like to inverse this equation and compute dp from deto. 
Many thanks for your help.
Cheers.
Ghiath

**Comment by thelfer** [comment]
Hi @GhiathMonnet,

Usually, `p` (the equivalent plastic strain) is an internal state variable, so I am not sure to understand your request.

What dsl are you using ?

**Comment by GhiathMonnet** [comment]
Yes. I agree. But I used the Implicit parser to implement a behavior with two branches : elstoplastic and viscoelastic. seq is the sum of the equivalent stress of the two branches. The equivalent plastic strain is thus different in the two braches. I need to compute the stress in the linear hardeing plastic branche to deduce the behavior of the viscoelastic brache. I found a very simple implementation which works perfectly, but I had to use a trick which is not satisfactory at all to deduce sigp (the stress in the plastic branches). 

  const auto eps11 = eto[0];
  const auto deps11 = deto[0];
  if (eps11 <= epsc) {   // epsc = the elastic strain at the yield stress) 
    dpp = 0 ;
    sigp = eps11 * Kp ;    // Kp = the young modulus of the plastic brache 
  } else {
    dpp = deps11 * (1. - tau/Kp );   // tau = the effective young modulus for the two branches
    sigp = sigc + pp * alpha ;
  }   
   
Here I considered the plastic strain in the plastic branche (pp) as the state variable. I am sure we can do much better ...

**Reply by thelfer** [reply]
I don't understand what you mean by `seq` is the sum of the equivalent stress of the two branches. If the total stress is the sum of two contributions, the equivalent stress is the sum of the equivalent stresses only if the stress in the two branches are proportional...

**Comment by GhiathMonnet** [comment]
Sorry : tau is the hardening modulus of the plastic branches

**Comment by GhiathMonnet** [comment]
I am not sur I understand your remarque. The two branches are independent. There is no relation between seq_v in the viscous branche and seq_p in the plastic branche, though we can still write :  seq = seq_v + seq_p.  The equivalent stress in every branche is :
seq_v = Dv * (eto - pv * n)  // pv : the equivalent viscous strain (dpv = A*pwo(seq_v,m)*dt 
seq_p = sigp = Dp * (eto - pp * n) // pp the eqivalent plasic strain
Is this clear ?

**Reply by thelfer** [reply]
To be clearer `seq_v = Dv * (eto - pv * n)` defines a tensor, not a scalar.
My remark simply states that adding the tensorial contributions of each branch seems to make sense. This does not seem the case for equivalent stresses.

**Comment by GhiathMonnet** [comment]
Yes you are right. It is a tensor. 
So, if i come back my initial question, is there a tool in mfront function galery to project the tensor (deto) on the flow direction tensor (n) ? I guess it would be enough to inverse the matrix n. Many thanks.
Cheers

**Reply by thelfer** [reply]
The projection is simply `(deto|n)` but that gives you a scalar. You may use `n^n` as a fourth order projector, but this is clearly not invertible...

**Comment by GhiathMonnet** [comment]
Great !! it needs to be scalar, becaus if deto = dp * n, then deto | n = dp is the desired scalar quantity. 
Many thanks Thomas. I will send the final mfront file for your expert investigation.
